SAM-2586 Add Side Panel to Track Question Progress

master-bob opened the pull request:
Adds a side panel in tests, which are a single question per page, which shows test takers their progress in the test. Shows test takers what question they are on, all the other questions, which questions are answered, which questions are unanswered, and which questions are marked for review. Also allows test takers to navigate using the panel.

![questionprogress_example](https://cloud.githubusercontent.com/assets/6335721/8290574/eb4ef9b2-18f1-11e5-9769-e74719ebbd00.png)

steveswinsburg commented:
The styling of the panel looks a bit odd, also those look like clickable radio buttons. Maybe need some UX input?  Cool feature though.

bjones86 commented on samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/bean/delivery/DeliveryBean.java:
```diff
@@ -809,6 +809,13 @@ public void setPrevious(boolean previous)
   //Settings
   public String getQuestionLayout()
   {
+      if(getSettings().isFormatByQuestion())
```
Braces on single line if/else statements please.

master-bob replied:
Yes sir!

master-bob replied:
Yes, you are right. I'll fix it.

bjones86 commented on samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/bean/delivery/DeliveryBean.java:
```diff
@@ -1859,41 +1911,45 @@ public String next_page()
+        log.debug(sae.getMessage());
+        return "takeAssessment";
+      }
     }
```
All the above looks like just whitespace changes. I know it's tempting, my OCD wants to do this all the time. Unfortunately it's a history killer and we need to avoid doing this.

master-bob replied:
See L1903. This is a creation of a new function goto_question() which shares similar functionality / structure to the next_page() function. I copied and pasted the next_page() function from wget-ing the raw file off github, hopefully this will fix the diff. Else, I'm not sure what else to do.

bjones86 replied:
A common solution is to extract the code that is shared between the two methods into a new convenience method that the other two will call. In this way, you avoid code duplication.

master-bob replied:
You are correct! I will fix it!

bjones86 commented on samigo/samigo-app/src/webapp/css/tool_sam.css:
```diff
@@ -1060,4 +1060,93 @@ ul.no-list {
+	min-height: 200px;
+	max-height:auto;
+	width: 295px;
+	background: #ddd;
```
Tabs vs. spaces here; inconsistent with the rest of the file. Please just use whatever crazy tabs/spaces (or combination there of) is already dominant in the file.

master-bob replied:
The file is internally inconsistent. And I can't tell at a glance which is more dominant, tabs or spaces.

bjones86 commented on samigo/samigo-app/src/webapp/js/questionProgress.js:
```diff
@@ -0,0 +1,135 @@
+<!-- Javascript for the Question Progress Panel-->
+
```
I would suggest namespacing all this JavaScript. Not entirely necessary, but certainly a best practise.

bjones86 commented on samigo/samigo-app/src/webapp/jsf/delivery/deliverAssessment.jsp:
```diff
@@ -46,12 +46,24 @@
+       
+	#delivPageWrapper
+	{
+	     height:1800px;
```
Tabs. vs spaces inconsistent with above

bjones86 commented on samigo/samigo-app/src/webapp/jsf/delivery/deliverAssessment.jsp:
```diff
@@ -46,12 +46,24 @@
-	  <script type="text/javascript" src="/samigo-app/js/saveForm.js"></script>	  	  
-      <script type="text/javascript">
+	  <script type="text/javascript" src="/samigo-app/js/saveForm.js"></script>
+          <script type="text/javascript">
```
This looks like more whitespace changes...

bjones86 commented on samigo/samigo-app/src/webapp/jsf/delivery/questionProgress.jspf:
```diff
@@ -0,0 +1,84 @@
+<div class="tier1">
+<h2>
+    <h:outputText value="Question Progress " />
+  </h2>
```
I think you have some extra whitespace at the beginning of this line.

master-bob replied:
Yes, you are right. Thanks!

bjones86 commented:
Minor code changes suggestions provided

bjones86 commented:
Whoops, pressed the wrong button!

master-bob commented:
@bjones86, I've addressed all your issues. Moving all the javascript to it's own namespace unfortunately broke the panel. In the meantime, I will close this PR and after I have fixed it I will reopen it.

master-bob commented:
Fixed what I broke.

master-bob commented:
I botched this, I should have reopened before force pushing the fixes. I can't reopen this, so I am making a new PR.  See #763
